index.bs

```diff
@@ -488,13 +490,238 @@ partial interface NeuralNetworkContext {
+        - *activations*: an optional sequence of {{RecurrentNetworkActivation}}. A pair of activation functions with the first function used for the update and reset gate, and the second used for the new gate. When not specified, it's default to the sigmoid ("sigmoid") and the hyperbolic tangent ("tanh") function respectively.
+
+    **Returns:** a sequence of {{Operand}}. The first element of the sequence is a 3-D tensor of shape [num_directions, batch_size, hidden_size], the cell output from the last time step of the network. Additionally, if *returnSequence* is set to true, the second element is the 4-D output tensor of shape [steps, num_directions, batch_size, hidden_size] containing every cell outputs from each time step in the temporal sequence.
+
+    <div class="note">
+    The behavior of this operation with default argument values can be generically emulated from the usage of other operations as follow. However, user agents typically have a more efficient implementation for it, therefore its usage is encouraged from the performance standpoint.
+    <pre highlight="js">
+    let hiddenState = initialHiddenState;
+    let slots = (direction == "both" ? 2 : 1);
+    let sequence = null;
+
+    let cellWeight = [];
+    let cellRecurrentWeight = [];
+    let cellBias = [];
+    let cellRecurrentBias = [];
+
+    for (let slot = 0; slot < slots; ++slot) {
+      cellWeight.push(nn.squeeze(nn.slice(weight, [slot, 0, 0], [slot + 1, -1, -1])), [0]);
+      cellRecurrentWeight.push(nn.squeeze(nn.slice(recurrentWeight, [slot, 0, 0], [slot + 1, -1, -1])), [0]);
+      cellBias.push(bias ? (nn.squeeze(nn.slice(bias, [slot, 0], [slot + 1, -1]), [0])) : null);
+      cellRecurrentBias.push(recurrentBias ? (nn.squeeze(nn.slice(recurrentBias, [slot, 0], [slot + 1, -1]), [0])) : null);
+    }
+
+    for (let step = 0; step < steps; ++step) {
+      let cellHidden = [];
+      let cellOutput = null;
+
+      for (let slot = 0; slot < slots; ++slot) {
+        cellHidden.push(nn.squeeze(nn.slice(hidden, [slot, 0, 0], [slot + 1, -1, -1]), [0]));
+      }
+
+      for (let slot = 0; slot < slots; ++slot) {
+        let slice = (slot == 1 || direction == "backward" ? steps - step - 1 : step);
+        let cellInput = nn.squeeze(nn.slice(input, [slice, 0, 0], [slice + 1, -1, -1]), [0]);
+
+        let result = nn.gruCell(
+          cellInput, cellWeight[slot], cellRecurrentWeight[slot], cellHidden[slot], 
+          hiddenSize, cellBias[slot], cellRecurrentBias[slot], resetAfter, layout, activations);
+
+        cellOutput = (cellOutput ? nn.concat([cellOutput, result], 0) : result);
+      }
+
+      hiddenState = cellOutput;
+
+      if (returnSequence) {
+        cellOutput = nn.reshape(cellOutput, [1, slots, -1, hiddenSize]);
+        sequence = (sequence ? nn.concat([sequence, cellOutput], 0) : cellOutput);
+      }
+    }
+
+    return (sequence ? [hiddenState, sequence] : [hiddenState]);
```

Comment:
nit: may be use "returnSequence ? [hiddenState, sequence] : [hiddenState]"? There will be a difference only in the case where steps=0.

Reply:
I guess that does raise another question: the returned value should probably be an empty tensor (of size 0) rather than null.

Reply:
steps == 0 should not be allowed. I'll add the clarification text on the argument.